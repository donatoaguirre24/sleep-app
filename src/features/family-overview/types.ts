export type SleepIntervalOverview = {
  intervalId: string
  userId: string
  name: string
  averageHeartRate: number
  averageRespiratoryRate: number
  score: number
  timeAsleep: number
}

export type DailyFamilyOverview = {
  date: string
  data: Array<SleepIntervalOverview>
}
