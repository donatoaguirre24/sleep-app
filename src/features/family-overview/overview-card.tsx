import { useTheme } from '@shopify/restyle'
import React, { useMemo } from 'react'
import { Pressable, PressableStateCallbackType } from 'react-native'

import { strings } from '@/locales'
import { getTimeFromSeconds } from '@/services'
import { Box, Card, CardProps, CircularProgress, Text, Theme } from '@/ui'
import { SleepIntervalOverview } from './types'

type Props = CardProps & {
  data: SleepIntervalOverview
  onSelection: (data: SleepIntervalOverview) => void
}

export const OverviewCard: React.FC<Props> = ({ data, onSelection, ...props }) => {
  const { colors } = useTheme<Theme>()
  const { hours, minutes } = getTimeFromSeconds(data.timeAsleep)

  const timeAsleep = `${hours}${strings.common.hoursShort} ${minutes}${strings.common.minutesShort} ${strings.common.asleep}`
  const averageBpm = `${data.averageHeartRate}${strings.common.bpm} (${strings.common.averageShort})`

  const scoreColor = useMemo(() => {
    if (data.score < 50) {
      return colors.red
    }
    if (data.score < 75) {
      return colors.orange
    }
    return colors.green
  }, [data.score, colors])

  const handlePress = () => {
    onSelection(data)
  }

  const computeStylesFromState = ({ pressed }: PressableStateCallbackType) => ({
    opacity: pressed ? 0.6 : 1,
    transform: [{ scale: pressed ? 0.98 : 1 }],
  })

  return (
    <Pressable onPress={handlePress} style={computeStylesFromState}>
      <Card
        backgroundColor="card"
        flexDirection="row"
        alignItems="center"
        variant="elevated"
        paddingVertical="m"
        {...props}
      >
        <CircularProgress
          flex={1}
          marginRight="s"
          color={scoreColor}
          percentage={data.score}
          size={65}
          strokeWidth={5}
        />
        <Box flex={3}>
          <Text variant="subheader" marginBottom="s">
            {data.name}
          </Text>
          <Box flexDirection="row">
            <Box flex={1}>
              <Text color="secondaryGray" fontWeight="600">
                {timeAsleep}
              </Text>
            </Box>
            <Box flex={1} paddingLeft="s">
              <Text color="magenta" fontWeight="600">
                â™¥ {averageBpm}
              </Text>
            </Box>
          </Box>
        </Box>
      </Card>
    </Pressable>
  )
}
