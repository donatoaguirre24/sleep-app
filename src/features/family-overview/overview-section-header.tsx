import { BoxProps } from '@shopify/restyle'
import React from 'react'

import { dateStringFromISODate } from '@/services'
import { Box, Text, Theme } from '@/ui'

type Props = BoxProps<Theme> & {
  date: string
}

export const OverviewSectionHeader: React.FC<Props> = ({ date, ...props }) => {
  const formattedDate = dateStringFromISODate(date)

  return (
    <Box bg="background" {...props}>
      <Text variant="subheader">{formattedDate}</Text>
    </Box>
  )
}
