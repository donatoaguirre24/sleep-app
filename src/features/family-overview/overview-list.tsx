import { useTheme } from '@shopify/restyle'
import React from 'react'
import { SectionList } from 'react-native'

import { Theme } from '@/ui'
import { OverviewCard } from './overview-card'
import { OverviewSectionHeader } from './overview-section-header'
import { DailyFamilyOverview, SleepIntervalOverview } from './types'

type Props = {
  data: Array<DailyFamilyOverview>
  onSelection: (data: SleepIntervalOverview) => void
}

export const OverviewList: React.FC<Props> = ({ data, onSelection }) => {
  const { spacing } = useTheme<Theme>()

  return (
    <SectionList
      contentContainerStyle={{ paddingBottom: spacing.m, paddingHorizontal: spacing.m }}
      contentInsetAdjustmentBehavior="automatic"
      keyExtractor={(item, index) => item.userId + index}
      sections={data}
      renderItem={({ item }) => (
        <OverviewCard marginBottom="m" data={item} onSelection={onSelection} />
      )}
      renderSectionHeader={({ section }) => (
        <OverviewSectionHeader paddingVertical="s" rawDate={section.date} />
      )}
    />
  )
}
