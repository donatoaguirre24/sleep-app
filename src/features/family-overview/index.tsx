import React, { useCallback } from 'react'
import { SafeAreaView } from 'react-native'

import data from '@/data/overview.json'
import { NavigationProps, Screens } from '@/navigation'
import { OverviewList } from './overview-list'

export const FamilyOverview: React.FC<NavigationProps> = ({ navigation }) => {
  const { navigate } = navigation

  const showSleepData = useCallback(
    (intervalId: string) => {
      navigate(Screens.SleepDetails, { intervalId })
    },
    [navigate]
  )

  return (
    <SafeAreaView>
      <OverviewList data={data} onSelection={showSleepData} />
    </SafeAreaView>
  )
}
