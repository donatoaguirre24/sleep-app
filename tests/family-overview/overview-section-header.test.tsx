import { render, screen } from '@testing-library/react-native'
import React from 'react'

import { OverviewSectionHeader } from '@/features/family-overview/overview-section-header'
import { withProviders } from '../utils'

describe('OverviewSectionHeader', () => {
  it('renders correctly', () => {
    render(withProviders(<OverviewSectionHeader date="2017-03-09T08:08:00.000Z" />))

    expect(screen.toJSON()).toMatchSnapshot()
  })

  it('should render the date properly formatted', () => {
    render(withProviders(<OverviewSectionHeader date="2017-03-09T08:08:00.000Z" />))

    const firstCard = screen.getByText('March 9, 2017')

    expect(firstCard).toBeDefined()
  })
})
