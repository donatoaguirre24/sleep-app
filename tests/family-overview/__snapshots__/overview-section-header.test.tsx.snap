// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OverviewSectionHeader renders correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#F2F2F2",
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#1C1C1E",
            "fontSize": 20,
            "fontWeight": "600",
            "lineHeight": 26,
          },
        ]
      }
    >
      March 9, 2017
    </Text>
  </View>
</RNCSafeAreaProvider>
`;
